src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -216,25 +214,61 @@ private void updateGridMode() {
     sendChannelFocus();
   }
 
+  // When turning a knob that sometimes controls Hue, sometimes Brightness,
+  // sometimes Saturation, look up the current mode of the surface and return
+  // the appropriate subparameter that this knob currently controls.
+  private LXListenableNormalizedParameter getActiveSubparameter(AggregateParameter agg) {
+    if (agg instanceof LinkedColorParameter) {
+      LinkedColorParameter lcp = (LinkedColorParameter) agg;
+      if (lcp.mode.getEnum() == LinkedColorParameter.Mode.PALETTE) {
+        // If it's palette-linked, the knob always controls the palette index.
+        return lcp.index;
+      }
+    }
+
+    // So far, ColorParameters are the only kind of AggregateParameter. If
+    // we add more later, we'll have to add custom code here to handle them.
+    ColorParameter colorParameter = (ColorParameter) agg;
```

Comment:
This should still check instanceof ColorParameter and return a null for subparameter if it doesn't recognize the AggregateParameter type... otherwise it'll throw an Exception and crash the whole thing.